### PR TITLE
TPU speed

### PR DESCRIPTION
**Pull request by VikParuchuri**

### CODE REVIEW

File: surya/common/surya/__init__.py [hunk 1]
```diff
@@ -580,4 +591,173 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
...
+                self.config.image_token_id,
+                device=inputs_embeds.device,
+                dtype=torch.long,
+            )
+            mask = input_ids == image_token_id_tensor
+            last_image_token_pos = (
+                mask.size(1)
+                - 1
+                - mask.flip(dims=[1]).long().argmax(dim=1, keepdim=True)
+            )
```
Comment (tarun-menta) [inline surya/common/surya/__init__.py:704]: Slightly confusing - Can't we just do `torch.argmax(dim=1, keepdim=True)`? I might be missing something though

File: surya/common/surya/encoder/__init__.py [hunk 2]
```diff
@@ -133,6 +139,133 @@ def apply_rotary_pos_emb_flashatt(
...
+                "removed and `position_embeddings` will be mandatory."
+            )
+            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
+            cos = emb.cos()
+            sin = emb.sin()
+        else:
+            cos, sin = position_embeddings
+        q, k = apply_rotary_pos_emb_vision(q, k, cos, sin)
+
+        attention_mask = torch.zeros([bsz, 1, seq_length, seq_length], dtype=torch.bool)
```
Comment (tarun-menta) [inline surya/common/surya/encoder/__init__.py:178]: Minor - I'm not sure if this will be a speed boost, but we don't need to recompute these seq lens on every layer. Might make sense to compute once in the `model.encode_images` function and pass in the attention mask

File: surya/common/surya/encoder/__init__.py [hunk 3]
```diff
@@ -133,6 +139,133 @@ def apply_rotary_pos_emb_flashatt(
...
+        if position_embeddings is not None:
+            cos, sin = position_embeddings
+            q, k = apply_rotary_pos_emb_vision(q, k, cos, sin)
+
+        # Single reshape to flash attention format [batch, num_heads, seq_len, head_dim]
+        q = q.transpose(1, 2)  # [bsz, num_heads, seq_len, head_dim]
+        k = k.transpose(1, 2)
+        v = v.transpose(1, 2)
+
+        total_seqlen = q.shape[2]
```
Comment (tarun-menta) [inline surya/common/surya/encoder/__init__.py:247]: Same comment from SDPA - Can we precompute and save time?

**Review by tarun-menta — COMMENTED** [review]
Assuming that we always set the cache length such that we never actually roll the image out of sequence, LGTM! 

Added minor comments which are more tied to clarity, and speed. Don't see any big errors though!

File: surya/common/surya/__init__.py (hunk 1 above)
Comment (VikParuchuri) [reply surya/common/surya/__init__.py:704]: argmax gives the first image token, but we need the last
